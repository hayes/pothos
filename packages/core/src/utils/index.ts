import {
  type DirectiveNode,
  type FieldNode,
  type GraphQLDirective,
  type GraphQLField,
  getArgumentValues,
} from 'graphql';
import { PothosSchemaError, PothosValidationError } from '../errors';
import { InputListRef } from '../refs/input-list';
import { ListRef } from '../refs/list';
import {
  type InputType,
  type InputTypeParam,
  type MaybePromise,
  type OutputType,
  type PartialResolveInfo,
  type PothosOutputFieldConfig,
  type SchemaTypes,
  type TypeParam,
  typeBrandKey,
} from '../types';

export * from './base64';
export * from './context-cache';
export * from './enums';
export * from './input';
export * from './params';
export * from './sort-classes';

export function assertNever(value: never): never {
  throw new TypeError(`Unexpected value: ${value}`);
}

export function assertArray(value: unknown): value is unknown[] {
  if (!Array.isArray(value)) {
    throw new PothosValidationError('List resolvers must return arrays');
  }

  return true;
}

export function isThenable(value: unknown): value is PromiseLike<unknown> {
  return !!(
    value &&
    (typeof value === 'object' || typeof value === 'function') &&
    typeof (value as Record<string, unknown>).then === 'function'
  );
}

export function verifyRef(ref: unknown) {
  if (ref === undefined) {
    throw new PothosSchemaError(`Received undefined as a type ref.

This is often caused by a circular import
If this ref is imported from a file that re-exports it (like index.ts)
you may be able to resolve this by importing it directly from the file that defines it.
`);
  }
}

export function verifyInterfaces(interfaces: unknown) {
  if (!interfaces || typeof interfaces === 'function') {
    return;
  }

  if (!Array.isArray(interfaces)) {
    throw new PothosSchemaError('interfaces must be an array or function');
  }

  for (const iface of interfaces) {
    if (iface === undefined) {
      throw new PothosSchemaError(`Received undefined in list of interfaces.

This is often caused by a circular import
If this ref is imported from a file that re-exports it (like index.ts)
you may be able to resolve this by importing it directly from the file that defines it.

Alternatively you can define interfaces with a function that will be lazily evaluated,
which may resolver issues with circular dependencies:

Example:
builder.objectType('MyObject', {
  interface: () => [Interface1, Interface2],
  ...
});
`);
    }
  }
}

type BrandWithTypeResult<T, Types extends SchemaTypes> = T extends Record<string, unknown>
  ? {
      [K in keyof T | typeof typeBrandKey]: K extends keyof T
        ? T[K]
        : OutputType<Types>
    }
  : T;

export function brandWithType<T, Types extends SchemaTypes>(
  val: T,
  type: OutputType<Types>,
): BrandWithTypeResult<T, Types> {
  if (typeof val !== 'object' || val === null) {
    return val as BrandWithTypeResult<T, Types>;
  }

  Object.defineProperty(val, typeBrandKey, {
    enumerable: false,
    value: type,
  });

  return val as BrandWithTypeResult<T, Types>;
}

export function getTypeBrand(val: unknown) {
  if (typeof val === 'object' && val !== null && typeBrandKey in val) {
    return (val as { [typeBrandKey]: OutputType<SchemaTypes> })[typeBrandKey];
  }

  return null;
}

export function unwrapListParam<Types extends SchemaTypes>(
  param: InputTypeParam<Types> | TypeParam<Types>,
): InputType<Types> | OutputType<Types> {
  if (Array.isArray(param)) {
    return unwrapListParam(param[0]);
  }

  if (param instanceof ListRef || param instanceof InputListRef) {
    return unwrapListParam(param.listType as TypeParam<Types>);
  }

  return param;
}

export function unwrapOutputListParam<Types extends SchemaTypes>(
  param: TypeParam<Types>,
): OutputType<Types> {
  if (Array.isArray(param)) {
    return unwrapOutputListParam(param[0]);
  }

  if (param instanceof ListRef) {
    return unwrapOutputListParam(param.listType as TypeParam<Types>);
  }

  return param;
}

export function unwrapInputListParam<Types extends SchemaTypes>(
  param: InputTypeParam<Types>,
): InputType<Types> {
  if (Array.isArray(param)) {
    return unwrapInputListParam(param[0]);
  }

  if (param instanceof InputListRef) {
    return unwrapInputListParam(param.listType as InputTypeParam<Types>);
  }

  return param;
}

/**
 * Helper for allowing plugins to fulfill the return of the `next` resolver, without paying the cost of the
 * Promise if not required.
 */
export function completeValue<T, R>(
  valOrPromise: PromiseLike<T> | T,
  onSuccess: (completedVal: T) => PromiseLike<R> | R,
  onError?: (errVal: unknown) => PromiseLike<R> | R,
): Promise<Awaited<R>> | Awaited<R> {
  if (isThenable(valOrPromise)) {
    return Promise.resolve(valOrPromise).then(onSuccess, onError) as Promise<Awaited<R>>;
  }
  // No need to handle onError, this should just be a try/catch inside the `onSuccess` block
  const result = onSuccess(valOrPromise);

  // If the result of the synchronous call is a promise like, convert to a promise
  // for consistency
  if (isThenable(result)) {
    return Promise.resolve(result);
  }
  return result as Awaited<R>;
}

export function getMappedArgumentValues(
  def: GraphQLDirective | GraphQLField<unknown, unknown>,
  node: DirectiveNode | FieldNode,
  context: object,
  info: PartialResolveInfo,
) {
  const args = getArgumentValues(def, node, info.variableValues);
  const mappers = def.extensions?.pothosArgMappers as
    | PothosOutputFieldConfig<SchemaTypes>['argMappers']
    | undefined;

  if (mappers && mappers.length > 0) {
    return reduceMaybeAsync(mappers, args, (acc, argMapper) => argMapper(acc, context, info));
  }

  return args;
}

export function reduceMaybeAsync<T, R>(
  items: T[],
  initialValue: R,
  fn: (value: R, item: T, i: number) => MaybePromise<R>,
): MaybePromise<R> {
  function next(value: R, i: number): MaybePromise<R> {
    if (i === items.length) {
      return value;
    }

    return completeValue(fn(value, items[i], i), (result) => {
      return result === null ? (null as R) : next(result, i + 1);
    });
  }

  return next(initialValue, 0);
}
