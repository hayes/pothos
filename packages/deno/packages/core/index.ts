// @ts-nocheck
import './types/global/index.ts';
import { SchemaBuilder as SchemaBuilderClass } from './builder.ts';
import { FieldBuilder as InternalFieldBuilder } from './fieldUtils/builder.ts';
import { InputFieldBuilder as InternalInputFieldBuilder } from './fieldUtils/input.ts';
import { InterfaceFieldBuilder as InternalInterfaceFieldBuilder } from './fieldUtils/interface.ts';
import { MutationFieldBuilder as InternalMutationFieldBuilder } from './fieldUtils/mutation.ts';
import { ObjectFieldBuilder as InternalObjectFieldBuilder } from './fieldUtils/object.ts';
import { QueryFieldBuilder as InternalQueryFieldBuilder } from './fieldUtils/query.ts';
import { RootFieldBuilder as InternalRootFieldBuilder } from './fieldUtils/root.ts';
import { SubscriptionFieldBuilder as InternalSubscriptionFieldBuilder } from './fieldUtils/subscription.ts';
import { BaseTypeRef as InternalBaseTypeRef } from './refs/base.ts';
import { EnumRef as InternalEnumRef } from './refs/enum.ts';
import { InputListRef as InternalInputListRef } from './refs/input-list.ts';
import { InputObjectRef as InternalInputObjectRef } from './refs/input-object.ts';
import { InterfaceRef as InternalInterfaceRef } from './refs/interface.ts';
import { ListRef as InternalListRef } from './refs/list.ts';
import { ObjectRef as InternalObjectRef } from './refs/object.ts';
import { ScalarRef as InternalScalarRef } from './refs/scalar.ts';
import { UnionRef as InternalUnionRef } from './refs/union.ts';
import type { AddVersionedDefaultsToBuilderOptions, FieldKind, InputTypeParam, NormalizeSchemeBuilderOptions, RootName, SchemaTypes, TypeParam, } from './types/index.ts';
export * from './errors.ts';
export * from './plugins/index.ts';
export * from './types/index.ts';
export * from './utils/index.ts';
const SchemaBuilder = SchemaBuilderClass as unknown as {
    registerPlugin: typeof SchemaBuilderClass.registerPlugin;
    allowPluginReRegistration: boolean;
    new <Types extends Partial<PothosSchemaTypes.UserSchemaTypes> = {}>(options: Types extends {
        Defaults: "v3";
    } ? AddVersionedDefaultsToBuilderOptions<PothosSchemaTypes.ExtendDefaultTypes<Types>, "v3"> : NormalizeSchemeBuilderOptions<PothosSchemaTypes.ExtendDefaultTypes<Types>>): PothosSchemaTypes.SchemaBuilder<PothosSchemaTypes.ExtendDefaultTypes<Types>>;
};
export default SchemaBuilder;
export const FieldBuilder = InternalFieldBuilder as new <Types extends SchemaTypes, ParentShape, Kind extends Exclude<FieldKind, RootName> = Exclude<FieldKind, RootName>>(builder: PothosSchemaTypes.SchemaBuilder<Types>, kind: FieldKind, graphqlKind: PothosSchemaTypes.PothosKindToGraphQLType[FieldKind]) => PothosSchemaTypes.FieldBuilder<Types, ParentShape, Kind>;
export type RootFieldBuilder<Types extends SchemaTypes, ParentShape, Kind extends FieldKind = FieldKind> = PothosSchemaTypes.RootFieldBuilder<Types, ParentShape, Kind>;
export const RootFieldBuilder = InternalRootFieldBuilder as new <Types extends SchemaTypes, ParentShape, Kind extends FieldKind = FieldKind>(builder: PothosSchemaTypes.SchemaBuilder<Types>, kind: FieldKind, graphqlKind: PothosSchemaTypes.PothosKindToGraphQLType[FieldKind]) => PothosSchemaTypes.RootFieldBuilder<Types, ParentShape, Kind>;
export type QueryFieldBuilder<Types extends SchemaTypes, ParentShape> = PothosSchemaTypes.QueryFieldBuilder<Types, ParentShape>;
export const QueryFieldBuilder = InternalQueryFieldBuilder as new <Types extends SchemaTypes, ParentShape>(builder: PothosSchemaTypes.SchemaBuilder<Types>) => PothosSchemaTypes.QueryFieldBuilder<Types, ParentShape>;
export type MutationFieldBuilder<Types extends SchemaTypes, ParentShape> = PothosSchemaTypes.MutationFieldBuilder<Types, ParentShape>;
export const MutationFieldBuilder = InternalMutationFieldBuilder as new <Types extends SchemaTypes, ParentShape>(builder: PothosSchemaTypes.SchemaBuilder<Types>) => PothosSchemaTypes.MutationFieldBuilder<Types, ParentShape>;
export type SubscriptionFieldBuilder<Types extends SchemaTypes, ParentShape> = PothosSchemaTypes.SubscriptionFieldBuilder<Types, ParentShape>;
export const SubscriptionFieldBuilder = InternalSubscriptionFieldBuilder as new <Types extends SchemaTypes, ParentShape>(builder: PothosSchemaTypes.SchemaBuilder<Types>) => PothosSchemaTypes.SubscriptionFieldBuilder<Types, ParentShape>;
export type ObjectFieldBuilder<Types extends SchemaTypes, ParentShape> = PothosSchemaTypes.ObjectFieldBuilder<Types, ParentShape>;
export const ObjectFieldBuilder = InternalObjectFieldBuilder as new <Types extends SchemaTypes, ParentShape>(builder: PothosSchemaTypes.SchemaBuilder<Types>) => PothosSchemaTypes.ObjectFieldBuilder<Types, ParentShape>;
export type InterfaceFieldBuilder<Types extends SchemaTypes, ParentShape> = PothosSchemaTypes.InterfaceFieldBuilder<Types, ParentShape>;
export const InterfaceFieldBuilder = InternalInterfaceFieldBuilder as new <Types extends SchemaTypes, ParentShape>(builder: PothosSchemaTypes.SchemaBuilder<Types>) => PothosSchemaTypes.InterfaceFieldBuilder<Types, ParentShape>;
export type InputFieldBuilder<Types extends SchemaTypes, Kind extends "Arg" | "InputObject" = "Arg" | "InputObject"> = PothosSchemaTypes.InputFieldBuilder<Types, Kind>;
export const InputFieldBuilder = InternalInputFieldBuilder as new <Types extends SchemaTypes, Kind extends "Arg" | "InputObject" = "Arg" | "InputObject">(builder: PothosSchemaTypes.SchemaBuilder<Types>, kind: Kind, typename: string) => PothosSchemaTypes.InputFieldBuilder<Types, Kind>;
export type BaseTypeRef<Types extends SchemaTypes, T> = PothosSchemaTypes.BaseTypeRef<Types, T>;
export const BaseTypeRef = InternalBaseTypeRef as new <Types extends SchemaTypes, T>(kind: "Enum" | "InputObject" | "Interface" | "Object" | "Scalar" | "Union", name: string) => PothosSchemaTypes.BaseTypeRef<Types, T>;
export type EnumRef<Types extends SchemaTypes, T, P = T> = PothosSchemaTypes.EnumRef<Types, T, P>;
export const EnumRef = InternalEnumRef as new <Types extends SchemaTypes, T, P = T>(name: string) => PothosSchemaTypes.EnumRef<Types, T, P>;
export type InputObjectRef<Types extends SchemaTypes, T> = PothosSchemaTypes.InputObjectRef<Types, T>;
export const InputObjectRef = InternalInputObjectRef as new <Types extends SchemaTypes, T>(name: string) => PothosSchemaTypes.InputObjectRef<Types, T>;
export type InputListRef<Types extends SchemaTypes, T> = PothosSchemaTypes.InputListRef<Types, T>;
export const InputListRef = InternalInputListRef as new <Types extends SchemaTypes, T>(listType: InputTypeParam<Types>, required: boolean) => PothosSchemaTypes.InputListRef<Types, T>;
export type InterfaceRef<Types extends SchemaTypes, T, P = T> = PothosSchemaTypes.InterfaceRef<Types, T, P>;
export const InterfaceRef = InternalInterfaceRef as new <Types extends SchemaTypes, T, P = T>(name: string) => PothosSchemaTypes.InterfaceRef<Types, T, P>;
export type ObjectRef<Types extends SchemaTypes, T, P = T> = PothosSchemaTypes.ObjectRef<Types, T, P>;
export const ObjectRef = InternalObjectRef as new <Types extends SchemaTypes, T, P = T>(name: string) => PothosSchemaTypes.ObjectRef<Types, T, P>;
export type ScalarRef<Types extends SchemaTypes, T, U, P = T> = PothosSchemaTypes.ScalarRef<Types, T, U, P>;
export const ScalarRef = InternalScalarRef as new <Types extends SchemaTypes, T, U, P = T>(name: string) => PothosSchemaTypes.ScalarRef<Types, T, U, P>;
export type UnionRef<Types extends SchemaTypes, T, P = T> = PothosSchemaTypes.UnionRef<Types, T, P>;
export const UnionRef = InternalUnionRef as new <Types extends SchemaTypes, T, P = T>(name: string) => PothosSchemaTypes.UnionRef<Types, T, P>;
export type ListRef<Types extends SchemaTypes, T, P = T> = PothosSchemaTypes.ListRef<Types, T, P>;
export const ListRef = InternalListRef as new <Types extends SchemaTypes, T, P = T>(listType: TypeParam<Types>, nullable: boolean) => PothosSchemaTypes.ListRef<Types, T, P>;
export { BuildCache } from './build-cache.ts';
export { ArgumentRef } from './refs/arg.ts';
export { BuiltinScalarRef } from './refs/builtin-scalar.ts';
export { FieldRef } from './refs/field.ts';
export { InputTypeRef } from './refs/input.ts';
export { InputFieldRef } from './refs/input-field.ts';
export { ImplementableInputObjectRef } from './refs/input-object.ts';
export { ImplementableInterfaceRef } from './refs/interface.ts';
export { MutationRef } from './refs/mutation.ts';
export { ImplementableObjectRef } from './refs/object.ts';
export { OutputTypeRef } from './refs/output.ts';
export { QueryRef } from './refs/query.ts';
export { SubscriptionRef } from './refs/subscription.ts';
