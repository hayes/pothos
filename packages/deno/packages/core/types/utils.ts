// @ts-nocheck
import type { GraphQLResolveInfo } from 'https://cdn.skypack.dev/graphql?dts';
export type MaybePromise<T> = Promise<T> | T;
// biome-ignore lint/suspicious/noExplicitAny: <explanation>
export type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never;
export type RequiredKeys<T extends object> = Exclude<keyof T, OptionalKeys<T>>;
export type OptionalKeys<T extends object> = {
    [K in keyof T]: T[K] | undefined extends T[K] ? K : T[K] | null extends T[K] ? K : never;
}[keyof T];
export type NonEmptyKeys<T extends object> = undefined extends {} ? never : {
    [K in keyof T]: {} extends T[K] ? never : T[K] extends NonNullable<T[K]> ? K : never;
}[keyof T];
export type EmptyKeys<T extends object> = {
    [K in keyof T]: {} extends T[K] ? K : T[K] extends NonNullable<T[K]> ? never : K;
}[keyof T];
export type Normalize<T> = T extends object ? {
    [K in keyof T]: T[K];
} : T;
export type NullableToOptional<T> = T extends object ? Normalize<{
    [K in OptionalKeys<T>]?: T[K];
} & {
    [K in RequiredKeys<T>]: T[K];
}> : T;
export type EmptyToOptional<T> = T extends object ? Normalize<{
    [K in EmptyKeys<T> | OptionalKeys<T>]?: T[K];
} & {
    [K in NonEmptyKeys<T>]: T[K];
}> : T;
export type NormalizeNullable<T> = undefined extends T ? T | null | undefined : null extends T ? T | null | undefined : T;
export type NormalizeNullableFields<T extends object> = Normalize<{
    [K in OptionalKeys<T>]?: T[K] | null | undefined;
} & {
    [K in RequiredKeys<T>]: T[K];
}>;
// Check if T is a Record of string keys who's values are not functions
export type IsSimpleRecord<T> = ([
    T
] extends [
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    Record<string, any>
] ? keyof T extends infer K ? K extends string ? T[K] extends (...args: any[]) => unknown ? [
    1
] extends [
    T[K]
] ? never : false : never : never : false : false) extends never ? true : false;
export type RecursivelyNormalizeNullableFields<T> = T extends null | undefined ? null | undefined : T extends (infer L)[] ? RecursivelyNormalizeNullableFields<L>[] : T extends (...args: any[]) => unknown ? T : keyof T extends string ? IsSimpleRecord<T> extends true ? Normalize<{
    [K in OptionalKeys<T & object>]?: K extends string ? RecursivelyNormalizeNullableFields<NonNullable<T[K]>> | null | undefined : T[K];
} & {
    [K in RequiredKeys<T & object>]: RecursivelyNormalizeNullableFields<NonNullable<T[K]>>;
}> : T : T;
export type RemoveNeverKeys<T extends {}> = {
    [K in keyof T as [
        T[K]
    ] extends [
        never
    ] ? never : K]: T[K];
};
export type Merge<T> = {
    [K in keyof T]: T[K];
} & {};
export type MergeUnion<T, Keys extends keyof T = T extends unknown ? keyof T : never> = Merge<T extends unknown ? {
    [K in Keys as K extends keyof T ? never : K]?: never;
} & {
    [K in Keys as K extends keyof T ? K : never]: T[K & keyof T];
} : never>;
export interface Path {
    prev: Path | undefined;
    key: number | string;
    typename: string | undefined;
}
export type LastIndex<T extends unknown[]> = T extends [
    unknown,
    ...infer U
] ? U["length"] : 0;
export type NormalizeArgs<T extends unknown[], Index extends keyof T = LastIndex<T>> = undefined extends T[Index] ? {} extends T[Index] ? undefined extends {} ? {
    [K in keyof T]?: T[K];
} : T : {
    [K in keyof T]-?: T[K];
} : {} extends T[Index] ? {
    [K in keyof T]?: T[K];
} : T;
export type IsStrictMode = undefined extends {} ? false : true;
export interface PartialResolveInfo {
    fragments: GraphQLResolveInfo["fragments"];
    variableValues: GraphQLResolveInfo["variableValues"];
    schema: GraphQLResolveInfo["schema"];
}
export type Simplify<T> = {
    [KeyType in keyof T]: T[KeyType];
} & {};
