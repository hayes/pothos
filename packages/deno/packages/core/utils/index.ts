// @ts-nocheck
import { type DirectiveNode, type FieldNode, type GraphQLDirective, type GraphQLField, getArgumentValues, } from 'https://cdn.skypack.dev/graphql?dts';
import { PothosSchemaError, PothosValidationError } from '../errors.ts';
import { InputListRef } from '../refs/input-list.ts';
import { ListRef } from '../refs/list.ts';
import { type InputType, type InputTypeParam, type OutputType, type PartialResolveInfo, type PothosOutputFieldConfig, type SchemaTypes, type TypeParam, typeBrandKey, } from '../types/index.ts';
export * from './base64.ts';
export * from './context-cache.ts';
export * from './enums.ts';
export * from './input.ts';
export * from './params.ts';
export * from './sort-classes.ts';
export function assertNever(value: never): never {
    throw new TypeError(`Unexpected value: ${value}`);
}
export function assertArray(value: unknown): value is unknown[] {
    if (!Array.isArray(value)) {
        throw new PothosValidationError("List resolvers must return arrays");
    }
    return true;
}
export function isThenable(value: unknown): value is PromiseLike<unknown> {
    return !!(value &&
        (typeof value === "object" || typeof value === "function") &&
        typeof (value as Record<string, unknown>).then === "function");
}
export function verifyRef(ref: unknown) {
    if (ref === undefined) {
        throw new PothosSchemaError(`Received undefined as a type ref.

This is often caused by a circular import
If this ref is imported from a file that re-exports it (like index.ts)
you may be able to resolve this by importing it directly from the file that defines it.
`);
    }
}
export function verifyInterfaces(interfaces: unknown) {
    if (!interfaces || typeof interfaces === "function") {
        return;
    }
    if (!Array.isArray(interfaces)) {
        throw new PothosSchemaError("interfaces must be an array or function");
    }
    for (const iface of interfaces) {
        if (iface === undefined) {
            throw new PothosSchemaError(`Received undefined in list of interfaces.

This is often caused by a circular import
If this ref is imported from a file that re-exports it (like index.ts)
you may be able to resolve this by importing it directly from the file that defines it.

Alternatively you can define interfaces with a function that will be lazily evaluated,
which may resolver issues with circular dependencies:

Example:
builder.objectType('MyObject', {
  interface: () => [Interface1, Interface2],
  ...
});
`);
        }
    }
}
export function brandWithType<Types extends SchemaTypes>(val: unknown, type: OutputType<Types>) {
    if (typeof val !== "object" || val === null) {
        return;
    }
    Object.defineProperty(val, typeBrandKey, {
        enumerable: false,
        value: type,
    });
}
export function getTypeBrand(val: unknown) {
    if (typeof val === "object" && val !== null && typeBrandKey in val) {
        return (val as {
            [typeBrandKey]: OutputType<SchemaTypes>;
        })[typeBrandKey];
    }
    return null;
}
export function unwrapListParam<Types extends SchemaTypes>(param: InputTypeParam<Types> | TypeParam<Types>): InputType<Types> | OutputType<Types> {
    if (Array.isArray(param)) {
        return unwrapListParam(param[0]);
    }
    if (param instanceof ListRef || param instanceof InputListRef) {
        return unwrapListParam(param.listType as TypeParam<Types>);
    }
    return param;
}
export function unwrapOutputListParam<Types extends SchemaTypes>(param: TypeParam<Types>): OutputType<Types> {
    if (Array.isArray(param)) {
        return unwrapOutputListParam(param[0]);
    }
    if (param instanceof ListRef) {
        return unwrapOutputListParam(param.listType as TypeParam<Types>);
    }
    return param;
}
export function unwrapInputListParam<Types extends SchemaTypes>(param: InputTypeParam<Types>): InputType<Types> {
    if (Array.isArray(param)) {
        return unwrapInputListParam(param[0]);
    }
    if (param instanceof InputListRef) {
        return unwrapInputListParam(param.listType as InputTypeParam<Types>);
    }
    return param;
}
/**
 * Helper for allowing plugins to fulfill the return of the `next` resolver, without paying the cost of the
 * Promise if not required.
 */
export function completeValue<T, R>(valOrPromise: PromiseLike<T> | T, onSuccess: (completedVal: T) => R, onError?: (errVal: unknown) => R): Promise<R> | R {
    if (isThenable(valOrPromise)) {
        return Promise.resolve(valOrPromise).then(onSuccess, onError);
    }
    // No need to handle onError, this should just be a try/catch inside the `onSuccess` block
    const result = onSuccess(valOrPromise);
    // If the result of the synchronous call is a promise like, convert to a promise
    // for consistency
    if (isThenable(result)) {
        return Promise.resolve(result);
    }
    return result;
}
export function getMappedArgumentValues(def: GraphQLDirective | GraphQLField<unknown, unknown>, node: DirectiveNode | FieldNode, context: object, info: PartialResolveInfo) {
    const args = getArgumentValues(def, node, info.variableValues);
    const mappers = def.extensions?.pothosArgMappers as PothosOutputFieldConfig<SchemaTypes>["argMappers"] | undefined;
    if (mappers && mappers.length > 0) {
        return mappers.reduce<Record<string, unknown>>((acc, argMapper) => argMapper(acc, context, info), args);
    }
    return args;
}
