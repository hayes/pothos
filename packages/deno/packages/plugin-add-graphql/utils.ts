// @ts-nocheck
import { type SchemaTypes, createContextCache } from '../core/index.ts';
import { type GraphQLNamedType, isEnumType, isInputObjectType, isInterfaceType, isObjectType, isScalarType, isUnionType, } from 'https://cdn.skypack.dev/graphql?dts';
export const referencedTypes = createContextCache(() => new Set<GraphQLNamedType>());
export function addTypeToSchema<Types extends SchemaTypes>(builder: PothosSchemaTypes.SchemaBuilder<Types>, type: GraphQLNamedType) {
    if (builder.configStore.hasConfig(type.name as never)) {
        return;
    }
    if (isObjectType(type)) {
        builder.addGraphQLObject(type);
    }
    else if (isInterfaceType(type)) {
        builder.addGraphQLInterface(type);
    }
    else if (isUnionType(type)) {
        builder.addGraphQLUnion(type);
    }
    else if (isEnumType(type)) {
        builder.addGraphQLEnum(type);
    }
    else if (isInputObjectType(type)) {
        builder.addGraphQLInput(type);
    }
    else if (isScalarType(type)) {
        builder.addScalarType(type.name as never, type);
    }
}
export function addReferencedType<Types extends SchemaTypes>(builder: PothosSchemaTypes.SchemaBuilder<Types>, type: GraphQLNamedType) {
    if (referencedTypes(builder).has(type)) {
        return;
    }
    builder.configStore.onPrepare(() => addTypeToSchema(builder, type));
}
