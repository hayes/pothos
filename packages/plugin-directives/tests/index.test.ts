import { envelop, useEngine, useSchema } from '@envelop/core';
import { useRateLimiter } from '@envelop/rate-limiter';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import SchemaBuilder from '@pothos/core';
import {
  type GraphQLEnumType,
  type GraphQLInputObjectType,
  type GraphQLObjectType,
  Kind,
  printSchema,
} from 'graphql';
import gql from 'graphql-tag';

import * as GraphQL from 'graphql';
import schema from './example/schema';

const getEnvelop = envelop({
  plugins: [
    useEngine(GraphQL),
    useSchema(schema),
    useRateLimiter({
      identifyFn: () => '1',
    }),
  ],
});

const { execute } = getEnvelop();

describe('extends example schema', () => {
  it('generates expected schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('generates expected schema with directives', () => {
    expect(printSchemaWithDirectives(schema)).toMatchSnapshot();
  });

  it('has expected directives in extensions', () => {
    const types = schema.getTypeMap();

    expect(types.Obj.extensions?.directives).toStrictEqual({ o: { foo: 123 } });
    expect(types.Query.extensions?.directives).toStrictEqual({
      o: { foo: 123 },
    });

    const testField = (types.Query as GraphQLObjectType).getFields().test;

    expect(testField.extensions?.directives).toStrictEqual({
      f: { foo: 123 },
      rateLimit: {
        max: 1,
        window: '5s',
        message: 'Too many requests, please try again in 5 seconds.',
      },
    });
    expect(testField.args[0].extensions?.directives).toStrictEqual({ a: { foo: 123 } });

    expect(types.IF.extensions?.directives).toStrictEqual({ i: { foo: 123 } });
    expect(types.UN.extensions?.directives).toStrictEqual({ u: { foo: 123 } });
    expect(types.EN.extensions?.directives).toStrictEqual({ e: { foo: 123 } });
    expect((types.EN as GraphQLEnumType).getValue('ONE')?.extensions?.directives).toStrictEqual({
      ev: { foo: 123 },
    });
    expect(types.Date.extensions?.directives).toStrictEqual({ s: { foo: 123 } });
    expect(types.In.extensions?.directives).toStrictEqual({ io: { foo: 123 } });
    expect(
      (types.In as GraphQLInputObjectType).getFields().test.extensions?.directives,
    ).toStrictEqual({ if: { foo: 123 } });
  });

  it('constructs the astNode with defaultValue for inputs and args', () => {
    const types = schema.getTypeMap();

    const testField = (types.Query as GraphQLObjectType).getFields().test;

    const arg1 = testField.args[0];
    const myInput = testField.args[1];
    const myOtherInput = testField.args[2];

    expect(arg1.astNode?.defaultValue).toBeUndefined();
    expect(myInput.astNode?.defaultValue).toBeUndefined();
    expect(myOtherInput.astNode?.defaultValue).toBeDefined();
    expect(myOtherInput.astNode?.defaultValue).toStrictEqual({
      kind: Kind.OBJECT,
      fields: [],
    });

    const MyInput = (types.MyInput as GraphQLInputObjectType).getFields();

    expect(MyInput.booleanWithDefault.astNode?.defaultValue).toStrictEqual({
      kind: Kind.BOOLEAN,
      value: false,
    });
    expect(MyInput.enumWithDefault.astNode?.defaultValue).toStrictEqual({
      kind: Kind.ENUM,
      value: 'TWO',
    });
    expect(MyInput.idWithDefault.astNode?.defaultValue).toStrictEqual({
      kind: Kind.INT,
      value: '123',
    });
    expect(MyInput.stringWithDefault.astNode?.defaultValue).toStrictEqual({
      kind: Kind.STRING,
      value: 'default string',
    });
    expect(MyInput.idsWithDefault.astNode?.defaultValue).toStrictEqual({
      kind: Kind.LIST,
      values: [
        { kind: Kind.INT, value: '123' },
        { kind: Kind.INT, value: '456' },
      ],
    });
  });

  it('gatsby format', () => {
    const builder = new SchemaBuilder<{
      Directives: {
        rateLimit: {
          locations: 'FIELD_DEFINITION' | 'OBJECT';
          args: {
            limit: number;
            duration: number;
          };
        };
      };
    }>({
      plugins: ['directives'],
    });

    builder.queryType({
      directives: {
        rateLimit: { limit: 5, duration: 60 },
      },
    });

    const types = builder.toSchema().getTypeMap();

    expect(types.Query.extensions?.directives).toStrictEqual([
      {
        name: 'rateLimit',
        args: { limit: 5, duration: 60 },
      },
    ]);
  });

  it('rate limits', async () => {
    const query = gql`
      query {
        test
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: () => ({}),
    });

    expect(result).toMatchInlineSnapshot(`
      {
        "data": {
          "test": "hi",
        },
      }
    `);

    const result2 = await execute({
      schema,
      document: query,
      contextValue: () => ({}),
    });

    expect(result2).toMatchInlineSnapshot(`
      {
        "data": {
          "test": null,
        },
        "errors": [
          [GraphQLError: Too many requests, please try again in 5 seconds.],
        ],
      }
    `);
  });
});
