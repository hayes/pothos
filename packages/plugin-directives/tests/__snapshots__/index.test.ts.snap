// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extends example schema > generates expected schema 1`] = `
"scalar Date

enum EN {
  ONE
  TWO
}

interface IF {
  field: String
}

input In {
  test: String
}

input MyInput {
  booleanWithDefault: Boolean = false
  enumWithDefault: EN = TWO
  id: ID!
  idWithDefault: ID = 123
  ids: [ID!]!
  idsWithDefault: [ID!] = [123, 456]
  stringWithDefault: String = "default string"
}

input MyOtherInput {
  booleanWithDefault: Boolean = false
}

type Obj {
  field: String
}

type Query {
  cacheControlPrivate: String
  cacheControlPublic: String
  test(arg1: String, myInput: MyInput, myOtherInput: MyOtherInput = {}): String
}

union UN = Obj"
`;

exports[`extends example schema > generates expected schema with directives 1`] = `
"schema {
  query: Query
}

scalar Date @s(foo: 123)

enum EN @e(foo: 123) {
  ONE @ev(foo: 123)
  TWO
}

interface IF @i(foo: 123) {
  field: String
}

input In @io(foo: 123) {
  test: String @if(foo: 123)
}

input MyInput {
  booleanWithDefault: Boolean = false
  enumWithDefault: EN = TWO
  id: ID!
  idWithDefault: ID = 123
  ids: [ID!]!
  idsWithDefault: [ID!] = [123, 456]
  stringWithDefault: String = "default string"
}

input MyOtherInput {
  booleanWithDefault: Boolean = false
}

type Obj @o(foo: 123) {
  field: String
}

type Query @o(foo: 123) {
  cacheControlPrivate: String @cacheControl(scope: "PRIVATE", maxAge: 100, inheritMaxAge: true)
  cacheControlPublic: String @cacheControl(scope: "PUBLIC", maxAge: 100, inheritMaxAge: true)
  test(arg1: String @a(foo: 123), myInput: MyInput, myOtherInput: MyOtherInput = {}): String @f(foo: 123) @rateLimit(max: 1, window: "5s", message: "Too many requests, please try again in 5 seconds.")
}

union UN @u(foo: 123) = Obj"
`;
